conf.c

```diff
@@ -93,168 +128,482 @@ chomp(char *str)
+	print_masquerade_settings();
+#endif
+}
+
+void initialize_all_configuration_settings(void)
+{
+	if(is_configuration_initialized != false)
+		/* This function shall only run once... */
+		return;
+
+	is_configuration_initialized = true;
+
+	SLIST_INIT(&config_head);
+
+	/* SMARTHOST */
+	initialize_configuration_setting(CONF_SMARTHOST, NULL, true, NULL);
+	/* PORT */
+	initialize_configuration_setting(CONF_PORT, NULL, true, SMTP_PORT_STRING);
+	/* and so on */
+	initialize_configuration_setting(CONF_ALIASES, NULL, true, DEFAULT_ALIASES_PATH);
+	initialize_configuration_setting(CONF_SPOOLDIR, NULL, true, DEFAULT_SPOOLDIR);
+	initialize_configuration_setting(CONF_AUTHPATH, NULL, true, NULL);
+	initialize_configuration_setting(CONF_CERTFILE, NULL, true, NULL);
+	initialize_configuration_setting(CONF_MAILNAME, NULL, true,  NULL);
+	initialize_configuration_setting(CONF_MASQUERADE, NULL, true, NULL);
+	initialize_configuration_setting(CONF_STARTTLS, NULL, false, NULL);
+	initialize_configuration_setting(CONF_FINGERPRINT, &check_fingerprint_configuration, true, NULL);
+	initialize_configuration_setting(CONF_TLS_OPP, NULL, false, NULL);
+	initialize_configuration_setting(CONF_SECURETRANSFER, NULL, false, NULL);
+	initialize_configuration_setting(CONF_DEFER, NULL, false, NULL);
+	initialize_configuration_setting(CONF_INSECURE, NULL, false, NULL);
+	initialize_configuration_setting(CONF_FULLBOUNCE, NULL, false, NULL);
+	initialize_configuration_setting(CONF_NULLCLIENT, NULL, false, NULL);
+}
+
+static void initialize_configuration_setting(
+		const char* identifier,
+		check_ptr_t func_ptr, bool needs_value,
+		const char* default_value
+		)
+{
+	struct config_item_t *item;
+
+	if(identifier == NULL)
+		return;
 
-	while (!feof(conf)) {
-		if (fgets(line, sizeof(line), conf) == NULL)
+	/* Check if the setting has already been initialized */
+	SLIST_FOREACH(item, &config_head, next_item)
+		if(!strcmp(identifier, item->identifier))
+			/* Already in list */
+			errlogx(EX_SOFTWARE, "Trying to initialize setting '%s' more than once", identifier);
+
+	item = calloc(1, sizeof(struct config_item_t));
+	if(item == NULL)
+		errlog(EX_OSERR,"Error allocating memory\n");
+		/* exit */
+
+	/* We use strdup() here so that we'll be able to properly clean up everything */
+	item->identifier = strdup(identifier);
+	item->checkFunction = func_ptr;
+	item->needs_value = needs_value;
+
+	if(default_value != NULL)
+		item->str_value = strdup(default_value);
+	else
+		item->str_value = NULL;
+
+	SLIST_INSERT_HEAD(&config_head, item, next_item);
+}
+
+int try_to_set_configuration_setting(char *identifier, char *value)
+{
+	struct config_item_t *item = NULL;
+	bool found = false;
+
+	SLIST_FOREACH(item, &config_head, next_item)
+		if(!strcmp(identifier, item->identifier))
+		{
+			found = true;
 			break;
-		lineno++;
+		}
 
-		chomp(line);
+	if(found)		/* item is pointing to the correct item now */
+	{
+		 /* Let's see if it requires a value */
+		if(item->needs_value)
```

Comment:
maybe better to call it "boolean_flag"?

Reply:
Not sure here to be honest.
the line
if (item->needs_value) is much clearer in my eyes than
if (item->boolean_flag)
because its name already tells everything it's about: A flag to see if the configuration setting requires a value to be provided or not. boolean_flag pretty much says nothing about what it is about.

Reply:
I read "needs_value" as "this configuration setting is required for a valid config", not "this configuration setting does not need a value to be in effect" (i.e. it is a boolean).

Reply:
Now I see what you mean by boolean flag in this case. SECURETRANSFER for example is a boolean flag, AUTHPATH isn't, right? So that way it could certainly be called that way.